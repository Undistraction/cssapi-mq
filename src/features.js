import { map, prop, compose } from 'ramda';
import camelcase from 'camelcase';

import dimensionsValueRenderer from './renderers/valueRenderers/dimensionsValueRenderer';
import resolutionValueRenderer from './renderers/valueRenderers/resolutionValueRenderer';
import aspectRatioValueRenderer from './renderers/valueRenderers/aspectRatioValueRenderer';
import colorValueRenderer from './renderers/valueRenderers/colorValueRenderer';
import monochromeValueRenderer from './renderers/valueRenderers/monochromeValueRenderer';

export const ORIENTATION = Object.freeze(['portrait', 'landscape']);
export const SCAN = Object.freeze(['interlace', 'progressive']);
export const GRID = Object.freeze([0, 1]);
export const UPDATE = Object.freeze(['none', 'slow', 'fast']);
export const OVERFLOW_BLOCK = Object.freeze([
  'none',
  'scroll',
  'optional-paged',
]);
export const OVERFLOW_INLINE = Object.freeze(['none', 'scroll']);
export const COLOR_GAMUT = Object.freeze(['srgb', 'p3', 'rec2020']);
export const DISPLAY_MODE = Object.freeze([
  'fullscreen',
  'standalone',
  'minimal-ui',
  'browser',
]);

export const LINEAR_FEATURES = [
  {
    name: 'orientation',
    validValues: ORIENTATION,
  },
  {
    name: 'scan',
    validValues: SCAN,
  },
  {
    name: 'grid',
    validValues: GRID,
    allowNoArgument: true,
  },
  {
    name: 'update',
    validValues: UPDATE,
    allowNoArgument: true,
  },
  {
    name: 'overflow-block',
    validValues: OVERFLOW_BLOCK,
  },
  {
    name: 'overflow-inline',
    validValues: OVERFLOW_INLINE,
  },
  {
    name: 'color-gamut',
    validValues: COLOR_GAMUT,
  },
  {
    name: 'display-mode',
    validValues: DISPLAY_MODE,
  },
];

export const RANGED_FEATURES = [
  {
    name: 'width',
    valueRenderer: dimensionsValueRenderer,
  },
  {
    name: 'height',
    valueRenderer: dimensionsValueRenderer,
  },
  {
    name: 'resolution',
    valueRenderer: resolutionValueRenderer,
  },
  {
    name: 'aspect-ratio',
    valueRenderer: aspectRatioValueRenderer,
  },
  {
    name: 'color',
    valueRenderer: colorValueRenderer,
    config: {
      allowNoArgument: true,
    },
  },
  {
    name: 'color-index',
    valueRenderer: colorValueRenderer,
    config: {
      allowNoArgument: true,
    },
  },
  {
    name: 'monochrome',
    valueRenderer: monochromeValueRenderer,
    config: {
      allowNoArgument: true,
    },
  },
];

export const rangedFeatureNames = map(compose(camelcase, prop('name')))(
  RANGED_FEATURES
);
