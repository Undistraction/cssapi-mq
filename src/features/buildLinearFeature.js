import { isUndefined } from 'ramda-adjunct'
import { compose, prop, objOf, both, always, ifElse } from 'ramda'
import { matchWithSuccessOrFailure } from 'folktale-validations'
import { renderFeature } from '../renderers/cssRenderers/queryRenderer'
import { throwAPILinearFeatureError } from '../errors2'
import validatorForLinearFeature from '../validations/validatorForLinearFeature'

export default (name, possibleValues, allowNoArgument = false) => value =>
  compose(
    matchWithSuccessOrFailure(
      compose(renderFeature(name), prop(`value`), prop(`value`)),
      compose(throwAPILinearFeatureError(name), prop(`value`))
    ),
    validatorForLinearFeature(possibleValues, !allowNoArgument),
    ifElse(
      both(isUndefined, always(allowNoArgument)),
      always({}),
      objOf(`value`)
    )
  )(value)
