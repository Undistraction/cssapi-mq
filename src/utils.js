import {
  __,
  inc,
  is,
  divide,
  contains,
  compose,
  map,
  zipObj,
  prop,
  sort,
  toPairs,
  reverse,
  findIndex,
  propEq,
  nth,
  subtract,
  when,
} from 'ramda';
import { css } from 'styled-components';
import { UNITS, MEDIA_PREFIX } from './const';

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const isBoolean = is(Boolean);
export const isNumber = is(Number);
export const isObject = is(Object);
export const isString = is(String);

export const ensureArray = when(isString, mediaTypes => [mediaTypes]);

export const appendUnit = (value, unit) => `${value}${unit}`;

export const ensureBreakpointOrder = (breakpoints, ...args) =>
  sort((a, b) => subtract(prop(a)(breakpoints), prop(b)(breakpoints)))(args);

export const toBreakpointArray = compose(
  map(zipObj(['name', 'value'])),
  toPairs
);

export const unitIsRemOrEm = contains(__, [UNITS.EM, UNITS.REM]);

export const orderByValue = compose(reverse, sort(prop('value')));

export const namePropEquals = propEq('name');

const findBreakpointIndex = (breakpoint, breakpointsArray) =>
  findIndex(namePropEquals(breakpoint))(breakpointsArray);

export const getUpperLimit = (breakpointsArray, breakpoint) => {
  const index = findBreakpointIndex(breakpoint, breakpointsArray);
  return compose(prop('name'), nth(inc(index)))(breakpointsArray);
};

export const buildQueryDefinition = (...elements) =>
  `${MEDIA_PREFIX} ${elements.join(' and ')}`;

export const buildQuery = (definition, content) => css`
  ${definition} {
    ${content};
  }
`;

export const buildFeature = (name, value) => `(${name}: ${value})`;

export const toOutput = (unit, baseFontSize, value) =>
  appendUnit(unitIsRemOrEm(unit) ? divide(value, baseFontSize) : value, unit);
