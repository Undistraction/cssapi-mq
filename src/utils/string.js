import { map, append, prepend, compose, join, of } from 'ramda'

export const joinWithNoSpace = join(``)
export const joinWithComma = join(`, `)

export const wrapWith = (a, b = a) =>
  compose(joinWithNoSpace, prepend(a), append(b), of)

export const wrapWithSquareBrackets = wrapWith(`[`, `]`)
export const wrapWithSingleQuotes = wrapWith(`'`)

// eslint-disable-next-line import/prefer-default-export
export const toCommaSeparatedList = values => values.join(`, `)
export const joinWithSpace = join(` `)

export const quoteAndJoinWithComma = compose(
  joinWithComma,
  map(wrapWithSingleQuotes)
)

export const toList = compose(wrapWithSquareBrackets, quoteAndJoinWithComma)
