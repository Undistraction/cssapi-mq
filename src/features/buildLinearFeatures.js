import { mergeAll, map } from 'ramda'
import camelcase from 'camelcase'
import { LINEAR_FEATURES } from '../features'

import buildLinearFeature from '../features/buildLinearFeature'

const toLinearFeatures = map(({ name, validValues, allowNoArgument }) => ({
  [camelcase(name)]: buildLinearFeature(name, validValues, allowNoArgument),
}))

export default () => mergeAll(toLinearFeatures(LINEAR_FEATURES))
