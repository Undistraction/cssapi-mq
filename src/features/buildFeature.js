import { isUndefined } from '../utils/value';
import { validateFeature } from '../validations';
import { renderFeature } from '../renderers/cssRenderers/queryRenderer';

export default (name, possibleValues, allowNoArgument = false) => value => {
  if (!(isUndefined(value) && allowNoArgument)) {
    validateFeature(name, value, possibleValues);
  }
  return renderFeature(name, value);
};
