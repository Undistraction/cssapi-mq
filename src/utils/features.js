import { mediaTypesAreValid, validateOrientation } from '../validations';
import { ensureArray, buildFeature } from '../utils';
import { throwError, invalidMediaTypeErrorMessage } from '../errors';

export const buildMediaType = defaultMediaType => (
  mediaTypes = [defaultMediaType]
) => {
  const mediaTypesArray = ensureArray(mediaTypes);
  if (!mediaTypesAreValid(mediaTypesArray))
    throwError(invalidMediaTypeErrorMessage(mediaTypesArray));
  return mediaTypesArray.join(', ');
};

export const buildOrientation = () => value => {
  validateOrientation(value);
  return buildFeature('orientation', value);
};
