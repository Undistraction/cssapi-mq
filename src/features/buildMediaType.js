import { when, append, prop, objOf, compose } from 'ramda'
import { ensureArray, isEmptyArray } from 'ramda-adjunct'
import { matchWithSuccessOrFailure } from 'folktale-validations'
import { throwAPIMediaTypeError } from '../errors2'
import validateAPIDefaultMediaTypeArgs from '../validations/validators/validateAPIDefaultMediaTypeArgs'
import { toCommaSeparatedList } from '../utils/string'

export default defaultMediaType => (mediaTypes = [defaultMediaType]) =>
  compose(
    matchWithSuccessOrFailure(
      compose(toCommaSeparatedList, prop(`mediaTypes`), prop(`value`)),
      compose(throwAPIMediaTypeError, prop(`value`))
    ),
    validateAPIDefaultMediaTypeArgs,
    objOf(`mediaTypes`),
    when(isEmptyArray, append(defaultMediaType)),
    ensureArray
  )(mediaTypes)
