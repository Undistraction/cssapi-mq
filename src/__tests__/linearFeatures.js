import { values } from 'ramda'
import camelcase from 'camelcase'
import { LINEAR_FEATURES } from '../features'
import featureValues from './testHelpers/featureValues'
import { mqWithValidBreakpointsForRange } from './testHelpers/data'
import cssSerialiser from './helpers/cssSerialiser'
import { toList } from '../utils/string'

expect.addSnapshotSerializer(cssSerialiser)

describe(`linear features`, () => {
  const testLinearFeature = (
    name,
    validValuesMap,
    invalidValues,
    { allowNoArgument = false } = {}
  ) => {
    const validValues = values(validValuesMap)
    const methodName = camelcase(name)
    describe(`${methodName}()`, () => {
      describe(`linear feature`, () => {
        if (allowNoArgument) {
          it(`doesn't throw if no argument is supplied`, () => {
            expect(() =>
              mqWithValidBreakpointsForRange(`width`)[methodName]()
            ).not.toThrow()
          })

          it(`returns a valueless ${name}`, () => {
            expect(
              mqWithValidBreakpointsForRange(`width`)[methodName]()
            ).toMatchSnapshot()
          })
        } else {
          it(`throws if no argument is supplied`, () => {
            expect(() => mqWithValidBreakpointsForRange(`width`)[methodName]())
              .toThrowMultiline(`
              [cssapi-mq] ${name}() Arguments missing required key(s): ['value']
            `)
          })

          it(`throws if 'undefined' is supplied`, () => {
            expect(() =>
              mqWithValidBreakpointsForRange(`width`)[methodName](undefined)
            ).toThrowMultiline(`
              [cssapi-mq] ${name}() Arguments missing required key(s): ['value']
            `)
          })
        }

        for (const value of invalidValues) {
          it(`throws if argument is '${value}'`, () => {
            expect(() =>
              mqWithValidBreakpointsForRange(`width`)[methodName](value)
            ).toThrowMultiline(`
              [cssapi-mq] ${name}() Arguments included invalid value(s)
                â€“ value: Value wasn't on the whitelist: ${toList(validValues)}`)
          })
        }
        for (const value of validValues) {
          it(`returns the supplied ${name} for '${value}'`, () => {
            expect(
              mqWithValidBreakpointsForRange(`width`)[methodName](value)
            ).toMatchSnapshot()
          })
        }
      })
    })
  }

  for (const feature of LINEAR_FEATURES) {
    const { name, validValues, allowNoArgument } = feature
    const { invalidValues } = featureValues(camelcase(name))
    testLinearFeature(name, validValues, invalidValues, { allowNoArgument })
  }
})
