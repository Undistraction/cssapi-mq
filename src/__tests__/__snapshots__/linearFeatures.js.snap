// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linear features colorGamut() linear feature returns the supplied color-gamut for 'p3' 1`] = `(color-gamut: p3)`;

exports[`linear features colorGamut() linear feature returns the supplied color-gamut for 'rec2020' 1`] = `(color-gamut: rec2020)`;

exports[`linear features colorGamut() linear feature returns the supplied color-gamut for 'srgb' 1`] = `(color-gamut: srgb)`;

exports[`linear features colorGamut() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] color-gamut() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['srgb', 'p3', 'rec2020']`;

exports[`linear features displayMode() linear feature returns the supplied display-mode for 'browser' 1`] = `(display-mode: browser)`;

exports[`linear features displayMode() linear feature returns the supplied display-mode for 'fullscreen' 1`] = `(display-mode: fullscreen)`;

exports[`linear features displayMode() linear feature returns the supplied display-mode for 'minimal-ui' 1`] = `(display-mode: minimal-ui)`;

exports[`linear features displayMode() linear feature returns the supplied display-mode for 'standalone' 1`] = `(display-mode: standalone)`;

exports[`linear features displayMode() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] display-mode() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['fullscreen', 'standalone', 'minimal-ui', 'browser']`;

exports[`linear features grid() linear feature returns a valueless grid 1`] = `(grid)`;

exports[`linear features grid() linear feature returns the supplied grid for '0' 1`] = `(grid: 0)`;

exports[`linear features grid() linear feature returns the supplied grid for '1' 1`] = `(grid: 1)`;

exports[`linear features orientation() linear feature returns the supplied orientation for 'landscape' 1`] = `(orientation: landscape)`;

exports[`linear features orientation() linear feature returns the supplied orientation for 'portrait' 1`] = `(orientation: portrait)`;

exports[`linear features orientation() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] orientation() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['portrait', 'landscape']`;

exports[`linear features overflowBlock() linear feature returns the supplied overflow-block for 'none' 1`] = `(overflow-block: none)`;

exports[`linear features overflowBlock() linear feature returns the supplied overflow-block for 'optional-paged' 1`] = `(overflow-block: optional-paged)`;

exports[`linear features overflowBlock() linear feature returns the supplied overflow-block for 'scroll' 1`] = `(overflow-block: scroll)`;

exports[`linear features overflowBlock() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] overflow-block() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['none', 'scroll', 'optional-paged']`;

exports[`linear features overflowInline() linear feature returns the supplied overflow-inline for 'none' 1`] = `(overflow-inline: none)`;

exports[`linear features overflowInline() linear feature returns the supplied overflow-inline for 'scroll' 1`] = `(overflow-inline: scroll)`;

exports[`linear features overflowInline() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] overflow-inline() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['none', 'scroll']`;

exports[`linear features scan() linear feature returns the supplied scan for 'interlace' 1`] = `(scan: interlace)`;

exports[`linear features scan() linear feature returns the supplied scan for 'progressive' 1`] = `(scan: progressive)`;

exports[`linear features scan() linear feature throws if no argument is supplied 1`] = `[cssapi-rhythm] scan() Arguments included invalid value(s) – value: Value wasn't on the whitelist: ['interlace', 'progressive']`;

exports[`linear features update() linear feature returns a valueless update 1`] = `(update)`;

exports[`linear features update() linear feature returns the supplied update for 'fast' 1`] = `(update: fast)`;

exports[`linear features update() linear feature returns the supplied update for 'none' 1`] = `(update: none)`;

exports[`linear features update() linear feature returns the supplied update for 'slow' 1`] = `(update: slow)`;
